src/parser/mod.rs

```diff
@@ -787,7 +788,10 @@ impl<'tokens> Parser<'tokens> {
         self.expect(Token::Ampersand, "`&` to start an immutable reference type")?;
         let owned = self.accept(Token::Owned);
         let shared = if owned { Sharedness::Owned } else { Sharedness::Shared };
-        Ok(Type::Reference(cst::Mutability::Immutable, shared))
+        match self.parse_type_application() {
+            Ok(application) => Ok(Type::TypeApplication(Box::new(Type::Reference(cst::Mutability::Immutable, shared)), vec![application])),
+            Err(_) => Ok(Type::Reference(cst::Mutability::Immutable, shared))
+        }
```

Comment:
- Need to update the `TODO` comment above
- Also, why do we only parse element types for immutable references? Let's add it to mutable references as well, and ideally have them share a method now to reduce duplicated code.